bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java

```diff
@@ -2597,11 +2678,14 @@ public void setExpandedState(Object elementOrTreePath, boolean expanded) {
 		if (checkBusy())
 			return;
 		Widget item = internalExpand(elementOrTreePath, false);
-		if (item instanceof Item) {
+
+		if ((item instanceof Item)) {
 			if (expanded) {
-				createChildren(item);
+				if (expandAndCreateChildren(item).length == 1) {
```

Comment:
I would expect the additional extension to happen after the "ordinary" extension has happended completely, i.e., after creating childen and setting the expand state has happened. First performing the recusive extension and then setting the expand state does not look to be the right order to me.

Reply:
I have changed the order in which both operations happen

Reply:
```java
if (item instanceof Item) {
	setExpanded((Item) item, expanded);
	if (expanded) {
		if (expandAndCreateChildren(item).length == 1) {
			expandPathOfSingleChildren(getChildren(item)[0], getAutoExpandOnSingleChildLevel());
		}
	}
}
```

What currently happens:
1) set the expanded state (don't actually expand)
**if the expanded state is expanded: <call expandAndCreateChildren>**
2) set the expanded state again *(doesn't do anything)*
3) create the children
4) return the amount of children
5) check: does my parent only have one child? If so, start the recursive routine that expands trivial paths and keeps track of how many levels it has recursed

Steps 1-4 expand the node and do not yet add any behaviour. Step 5 is the addition - just as you had described in your comment.
From what I understood in your comment, your issue boils down to 1) and 2) being duplicates of eachother. A rather obvious solution would be to check for the expanded state so that I only set it once

```java
if (!expanded) {
  setExpanded((Item) item, false);
} else {
  if (expandAndCreateChildren(item).length == 1) {
    expandPathOfSingleChildren(getChildren(item)[0], getAutoExpandOnSingleChildLevel());
  }
}
```

I am a bit hesitant about making this change as it introduces additional control flow and complexity. However, `setExpanded` performs a call to winapi (... and probably to cocoa and gtk too), hence I can understand why we would want to limit it's use.